File: frontend/src/services/orderService.ts
import axiosClient from './axiosClient'
import type {
  CreateOrderRequest, Order, OrderSearchParams, OrderSummary, UpdateOrderStatusRequest,
} from '../types/order.types'
import type { CheckoutRequest } from '../types/cart.types'
import type { ApiResponse, PageResponse } from '../types/common.types'

const BASE = '/orders'

export const orderService = {
  // GET /orders/search (not the plain list endpoint) is used for every list fetch - keyword/
  // status/milkType/date-range/customerId (admin-only) filters all live there. Non-admin callers
  // are scoped server-side to their own orders regardless of the customerId param.
  search: (params: OrderSearchParams) =>
    axiosClient.get<ApiResponse<PageResponse<Order>>>(`${BASE}/search`, {
      params: { page: 0, size: 20, sort: 'orderDate,desc', ...params },
    }),

  getById: (id: string) =>
    axiosClient.get<ApiResponse<Order>>(`${BASE}/${id}`),

  // SUPER_ADMIN/FARM_MANAGER/DELIVERY_MANAGER only - pendingOrders (platform-wide, orders with no
  // delivery assignment yet, since Order.status only leaves PENDING once one is created - see
  // DeliveryEventConsumer in order-service) backs the Delivery Dashboard's "Orders Waiting for
  // Assignment" KPI. Reused as-is, not re-derived from delivery-service's own data.
  getSummary: () =>
    axiosClient.get<ApiResponse<OrderSummary>>(`${BASE}/summary`),

  // Used by the dashboard's recent-orders list, one request, not per row.
  getRecent: (size = 5) =>
    axiosClient.get<ApiResponse<PageResponse<Order>>>(`${BASE}/search`, {
      params: { page: 0, size, sort: 'orderDate,desc' },
    }),

  // Used by Customer Details, one request per view, not per row.
  getByCustomer: (customerId: string, size = 5) =>
    axiosClient.get<ApiResponse<PageResponse<Order>>>(BASE, {
      params: { customerId, page: 0, size, sort: 'orderDate,desc' },
    }),

  // GET /orders/subscription/{subscriptionId} - every order the daily subscription-order job has
  // generated for one subscription. Used by Subscription Details, one request, not per row.
  // Ownership-checked server-side: a non-owner gets an empty page, not an error.
  getBySubscription: (subscriptionId: string, size = 10) =>
    axiosClient.get<ApiResponse<PageResponse<Order>>>(`${BASE}/subscription/${subscriptionId}`, {
      params: { page: 0, size },
    }),

  create: (data: CreateOrderRequest) =>
    axiosClient.post<ApiResponse<Order>>(BASE, data),

  // Creates a ONE_TIME order from the caller's own server-side Cart - see order-service's
  // OrderServiceImpl.checkout(). Items/prices are never sent from here; the backend sources them
  // from the cart itself and clears it on success.
  checkout: (data: CheckoutRequest) =>
    axiosClient.post<ApiResponse<Order>>(`${BASE}/checkout`, data),

  updateStatus: (id: string, data: UpdateOrderStatusRequest) =>
    axiosClient.patch<ApiResponse<Order>>(`${BASE}/${id}/status`, data),

  cancel: (id: string) =>
    axiosClient.delete<ApiResponse<void>>(`${BASE}/${id}`),

  export: async (params: OrderSearchParams & { format: 'CSV' | 'EXCEL' | 'PDF' }) => {
    const res = await axiosClient.get(`${BASE}/export`, { params, responseType: 'blob' })
    const disposition = res.headers['content-disposition'] as string | undefined
    const filename = disposition?.match(/filename="?([^"]+)"?/)?.[1] ?? `orders.${params.format.toLowerCase()}`
    return { blob: res.data as Blob, filename }
  },
}

File: backend/order-service/src/main/java/com/farm2home/order/service/OrderService.java
package com.farm2home.order.service;

import com.farm2home.common.core.dashboard.OrderSummaryResponse;
import com.farm2home.common.core.reports.ReportPage;
import com.farm2home.common.core.reports.SalesReportRow;
import com.farm2home.common.core.reports.SalesReportSummary;
import com.farm2home.common.export.ExportFormat;
import com.farm2home.order.domain.enums.MilkType;
import com.farm2home.order.domain.enums.OrderStatus;
import com.farm2home.order.dto.request.CheckoutRequest;
import com.farm2home.order.dto.request.CreateOrderRequest;
import com.farm2home.order.dto.request.UpdateOrderStatusRequest;
import com.farm2home.order.dto.response.GenerationResultResponse;
import com.farm2home.order.dto.response.OrderResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.io.IOException;
import java.io.OutputStream;
import java.time.LocalDate;
import java.util.UUID;

public interface OrderService {

    /** Create a manual ONE_TIME order. */
    OrderResponse createManualOrder(CreateOrderRequest request, UUID customerId);

    /** Create a ONE_TIME order from the caller's own server-side Cart - see
     *  OrderServiceImpl.checkout() for why items are never taken from the request itself. */
    OrderResponse checkout(CheckoutRequest request, UUID customerId);

    /** List orders. {@code null} customerId returns all (admin). */
    Page<OrderResponse> findAll(UUID customerId, Pageable pageable);

    /** Get order by ID. Ownership enforced unless customerId is null (admin). */
    OrderResponse findById(UUID id, UUID customerId);

    /** Get all orders for a specific subscription. Ownership enforced unless customerId is null
     *  (admin) - same convention as findById: a non-owner gets an empty page, not an error,
     *  matching how findAll/search already stay silent about what a mismatched filter excludes. */
    Page<OrderResponse> findBySubscription(UUID subscriptionId, UUID customerId, Pageable pageable);

    /**
     * Transition order status. Validates allowed transitions. This is the admin/staff-facing
     * manual path - a DELIVERY_PARTNER's real workflow is delivery-service's own
     * PATCH /delivery/assignments/{id}/status (correctly ownership-scoped there); this order's
     * status is kept in sync with that automatically via {@link com.farm2home.order.kafka.DeliveryEventConsumer},
     * not by a delivery partner calling this method directly.
     * actorId identifies who performed the change for audit purposes - customerId is null for
     * admin calls (it means "don't scope the lookup to one customer"), so actorId is passed
     * separately rather than reused for that.
     */
    OrderResponse updateStatus(UUID id, UpdateOrderStatusRequest request, UUID customerId, boolean isAdmin,
            UUID actorId);

    /** Soft-cancel an order. See updateStatus for why actorId is separate from customerId. */
    void cancel(UUID id, UUID customerId, boolean isAdmin, UUID actorId);

    /** Get order summary metrics for the dashboard. */
    OrderSummaryResponse getSummary();

    /** Sales Report: filtered, paginated orders plus summary totals over the same filter set. */
    ReportPage<SalesReportRow, SalesReportSummary> getReport(LocalDate dateFrom, LocalDate dateTo,
            OrderStatus status, UUID customerId, MilkType milkType, Pageable pageable);

    /** Enterprise search: keyword (order number/notes) + status + date range + product filters.
     *  {@code null} customerId returns across all customers (admin), same convention as findAll. */
    Page<OrderResponse> search(UUID customerId, String keyword, LocalDate dateFrom, LocalDate dateTo,
            OrderStatus status, MilkType milkType, Pageable pageable);

    /** Streams orders matching the same filters as {@link #search} to a downloadable file. See
     *  OrderServiceImpl for why this isn't @Transactional. */
    void export(ExportFormat format, OutputStream out, UUID customerId, String keyword, LocalDate dateFrom,
            LocalDate dateTo, OrderStatus status, MilkType milkType, String sortBy, boolean ascending) throws IOException;
}
